packages/js/core/src/types/WrapError.ts

```diff
@@ -0,0 +1,334 @@
+  CLIENT_LOAD_WRAPPER,
+  CLIENT_GET_FILE,
+  CLIENT_GET_IMPLEMENTATIONS,
+  CLIENT_VALIDATE_RESOLUTION,
+  CLIENT_VALIDATE_ABI,
+  CLIENT_VALIDATE_RECURSIVE,
+  URI_RESOLUTION = 25,
+  URI_RESOLVER,
+  URI_NOT_FOUND,
+  WASM_INVOKE_ABORTED = 50,
+  WASM_SUBINVOKE_ABORTED,
+  WASM_INVOKE_FAIL,
+  WASM_MODULE_NOT_FOUND,
+  WASM_STATE_ERROR,
+  PLUGIN_INVOKE_FAIL = 75,
+  PLUGIN_METHOD_NOT_FOUND,
+  PLUGIN_ARGS_MALFORMED,
+}
+
+export type WrapErrorSource = Readonly<{
+  file?: string;
+  row?: number;
+  col?: number;
+}>;
+
+export interface WrapErrorOptions {
+  code: WrapErrorCode;
+  uri: string;
+  method?: string;
+  args?: string;
+  source?: WrapErrorSource;
+  resolutionStack?: CleanResolutionStep;
+  cause?: unknown;
+  prev?: Error;
+  stack?: string;
+}
+
+type RegExpGroups<T extends string> =
+  | (RegExpExecArray & {
+      groups?: { [name in T]: string | undefined } | { [key: string]: string };
+    })
+  | null;
+
+export class WrapError extends Error {
+  readonly name: string;
+  readonly code: WrapErrorCode;
+  readonly reason: string;
+  readonly uri: string;
+  readonly method?: string;
+  readonly args?: string;
+  readonly source?: WrapErrorSource;
+  readonly resolutionStack?: CleanResolutionStep;
+  readonly cause?: unknown;
+  readonly prev?: Error;
+
+  private static re = new RegExp(
+    [
+      // [A-z]+Error can be replaced with specific error names when finalized
+      /^(?:[A-z_: ]*; )?[A-z]+Error: [\w ]+\./.source,
+      // there is some padding added to the number of words expected in an error code
+      /(?:\r\n|\r|\n)code: (?<code>1?[0-9]{1,2}|2[0-4][0-9]|25[0-5]) (?:[A-Z]+ ?){1,5}/
+        .source,
+      /(?:\r\n|\r|\n)reason: (?<reason>(?:.|\r\n|\r|\n)*)/.source,
+      /(?:\r\n|\r|\n)uri: (?<uri>wrap:\/\/[A-z0-9_-]+\/.+)/.source,
+      /(?:(?:\r\n|\r|\n)method: (?<method>([A-z_]{1}[A-z0-9_]*)))?/.source,
+      /(?:(?:\r\n|\r|\n)args: (?<args>\{(?:.|\r\n|\r|\n)+} ))?/.source,
+      /(?:(?:\r\n|\r|\n)source: \{ file: "(?<file>.+)", row: (?<row>[0-9]+), col: (?<col>[0-9]+) })?/
+        .source,
+      /(?:(?:\r\n|\r|\n)uriResolutionStack: (?<resolutionStack>\[(?:.|\r\n|\r|\n)+]))?/
+        .source,
+      /(?:(?:\r\n|\r|\n){2}This exception was caused by the following exception:(?:\r\n|\r|\n)(?<cause>(?:.|\r\n|\r|\n)+))?$/
+        .source,
+    ].join("")
+  );
+
+  constructor(reason = "Encountered an exception.", options: WrapErrorOptions) {
+    super(WrapError.stringify(reason, options));
+    this.name = WrapError.codeToName(options.code);
+    this.code = options.code;
+    this.reason = reason;
+    this.uri = options.uri;
+    this.method = options.method;
+    this.args = options.args;
+    this.source = options.source;
+    this.resolutionStack = options.resolutionStack;
+    this.cause = options.cause;
+    this.stack = options.stack;
+    this.prev = options.prev;
+    Object.setPrototypeOf(this, WrapError.prototype);
+  }
+
+  static parse(error: string): WrapError | undefined {
```

Comment:
This is a very good point. I moved it to `src/packages/js/wasm/serc/helpers/wrapErrorUtils`.

Reply:
> It's only relevant for wasm wrappers

Is this true though? Can't this be used to parse any stringified `WrapError`? If not, I think that's a problem. To me, we should have a `WrapError` structure, which can be easily stringified in a human readable way, and parsed back into the structure itself.

Another way of thinking about this: why are we parsing a human-readable output of the error, and not an easier format like a json string or a msgpack byte array? I think that, if we're transmitting the whole `WrapError` structure over a boundary, we should encoded it in a way that doesn't require custom regex parsing to get the `WrapError` struct back. We might need to change some wasm interfaces to do this properly (i.e. change `string` to `bytes`). This could be something we do in the future, but good to keep in mind.

Reply:
Right, it can parse any stringified `WrapError`. Right now it's used for Wasm wrappers because error messages are written to the memory buffer as strings. The error messages may or may not be WrapErrors, depending on the situation.

Currently, errors are essentially strings that can be nested within each other. The message part of the error can be an entire other error string. This occurs when an error is thrown during a subinvocation and is either not handled or is re-thrown. When a message is only thrown in the top-level invocation, the message is not a stringified error.

```typescript
// subinvocation in assemblyscript
const result = My_Module.method(...);
if (!result.ok) {
  // the entire error is now the message part of a new error.
  // this also happens when unwrapping a bad result.
  throw new Error(result.error);
}
```

Parsing the strings makes it easy to maintain and return a clean error structure that prints a clean error message.

Reply:
It would be cool to do this with msgpack serialization in the future, 100%

Reply:
I know I explained stuff you already know, but I was just clarifying that we are currently only parsing the errors in `wasm-js`. Maybe end-users will want to be able to parse errors though.

Reply:
I moved the parsing logic back to `WrapError`